### PR TITLE
Logger support in Lite

### PR DESCRIPTION
**Pull request by lightningforever**
## What does this PR do?

This PR adds the logger base class and TensorBoardLogger implementation to Fabric.
```py
logger = TensorBoardLogger()
fabric = LightningLite(loggers=logger)

fabric.log("accuracy", acc)
fabric.log_dict({"accuracy": acc, "loss": loss})
```

The base class interface is mostly the same as in PyTorch Lightning, but with these key differences:

- Removed model checkpoint callback dependency and `after_save_checkpoint` hook. Lite won't have the same checkpointing mechanism through callbacks, and thus we will have to rework the way loggers access checkpoints.
- Removed `DummyLogger` (only needed for Trainer)
- Moved properties to the top (code organization)
- Property: `save_dir` -> `root_dir`. We are standardizing the way the log dir is defined, to be compatible with as many logger frameworks as possible. More info in #14188. 

The TensorBoard implementation changes very little. Trainer-only features and dependencies were removed:

- Removed hparams saving
- Removed OmegaConf dependency
- `save_dir` -> `root_dir`
- Removed `log_graph` argument in constructor (logging graph is still supported through `logger.log_graph()`)

**Future Work:**
- Support for `self.log` in LightningModule will be added in #16135

### Does your PR introduce any breaking changes? If yes, please list them.

No

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: src/lightning_fabric/utilities/logger.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
+# Copyright The PyTorch Lightning team.
```
Comment (lightningforever) [inline src/lightning_fabric/utilities/logger.py:1]: These util functions were moved without modification

File: src/pytorch_lightning/utilities/logger.py [hunk 2]
```diff
@@ -13,148 +13,12 @@
...
-
-import numpy as np
-import torch
-from torch import Tensor
+from typing import List, Tuple
 
 from pytorch_lightning.callbacks import Checkpoint
 
 
-def _convert_params(params: Union[Dict[str, Any], Namespace]) -> Dict[str, Any]:
```
Comment (lightningforever) [inline src/pytorch_lightning/utilities/logger.py:27]: These util functions were moved to lightning_fabric/utilities/loggers.py

File: src/lightning_fabric/loggers/logger.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        """Return the experiment version."""
+
+    @property
+    def root_dir(self) -> Optional[str]:
+        """Return the root directory where all versions of an experiment get saved, or `None` if the logger does
+        not save data locally."""
+        return None
+
+    @property
+    def log_dir(self) -> Optional[str]:
```
Comment (lightningforever) [inline src/lightning_fabric/loggers/logger.py:47]: Note: This will be what we initially proposed as "experiment_dir" in #14188

File: src/lightning_fabric/loggers/logger.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+
+    @abstractmethod
+    def log_metrics(self, metrics: Dict[str, float], step: Optional[int] = None) -> None:
+        """Records metrics. This method logs metrics as soon as it received them.
+
+        Args:
+            metrics: Dictionary with metric names as keys and measured quantities as values
+            step: Step number at which the metrics should be recorded
+        """
+        pass
```
Comment (lightningforever) [inline src/lightning_fabric/loggers/logger.py:65]: This should probably become `log_dict`, and another method `log`, to align with the rest Lightning's naming.

File: src/lightning_fabric/loggers/tensorboard.py [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+        logger = TensorBoardLogger("path/to/logs/rot", name="my_model")
+        logger.log_hyperparams({"epochs": 5, "optimizer": "Adam"})
+        logger.log_metrics({"acc": 0.75})
+        logger.finalize("success")
+    """
+    LOGGER_JOIN_CHAR = "-"
+
+    def __init__(
+        self,
+        root_dir: _PATH,
```
Comment (justusschock) [inline src/lightning_fabric/loggers/tensorboard.py:84]: since we don't have the default root dir from trainer here, should we combine `root_dir`, `name` and `sub_dir` to a single path?

**Review by justusschock — APPROVED** [review]

File: src/lightning_fabric/loggers/logger.py (hunk 4 above)
Comment (carmocca) [reply src/lightning_fabric/loggers/logger.py:65]: I don't agree. What woudl be the difference between `log` and `log_dict` for this class specifically?

`log_metrics` is explicit and avoids collision with a polymorphic `log()` method if one of the implementations support it

File: src/lightning_fabric/loggers/tensorboard.py [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+_TENSORBOARDX_AVAILABLE = RequirementCache("tensorboardX")
+if TYPE_CHECKING:
+    # assumes at least one will be installed when type checking
+    if _TENSORBOARD_AVAILABLE:
+        from torch.utils.tensorboard import SummaryWriter
+    else:
+        from tensorboardX import SummaryWriter  # type: ignore[no-redef]
+
+
+class TensorBoardLogger(Logger):
```
Comment (carmocca) [inline src/lightning_fabric/loggers/tensorboard.py:44]: Don't we want the `pytorch_lightning.loggers.tensorboard` module to import this class? Without deprecations

File: src/lightning_fabric/loggers/logger.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+from functools import wraps
+from typing import Any, Callable, Dict, Optional, Union
+
+from torch import Tensor
+from torch.nn import Module
+
+from lightning_fabric.utilities.rank_zero import rank_zero_only
+
+
+class Logger(ABC):
```
Comment (carmocca) [inline src/lightning_fabric/loggers/logger.py:27]: Same comment about importing this from PL

File: src/lightning_fabric/fabric.py [hunk 8]
```diff
@@ -148,6 +157,16 @@ def is_global_zero(self) -> bool:
         """Whether this rank is rank zero."""
         return self._strategy.is_global_zero
 
+    @property
+    def loggers(self) -> List[Logger]:
+        """Returns all loggers passed to Fabric."""
+        return self._loggers
+
+    @property
+    def logger(self) -> Logger:
```
Comment (Borda) [inline src/lightning_fabric/fabric.py:166]: shall this have index as an argument so not to be a property?

File: src/lightning_fabric/loggers/logger.py (hunk 4 above)
Comment (lightningforever) [reply src/lightning_fabric/loggers/logger.py:65]: It depends on what product wants the logger interaction and experience to be. Will a user interact mainly with the fabric object and its log methods? Or will a user mainly interact with the logger object, i.e., `logger.log_()*`, `logger.experiment.xyz()`? 
Undoubtedly, power users will want to log media datatypes and not just metrics. I don't see how that would all be enabled through the fabric object. And then users would write code with a mix of `fabric.log` and `logger.log_metrics` and `logger.experiment.log_image`.

File: src/lightning_fabric/loggers/tensorboard.py (hunk 5 above)
Comment (lightningforever) [reply src/lightning_fabric/loggers/tensorboard.py:84]: I'm not sure we want to deviate that strongly from the design yet. I believe it's too early to tell. Eventually this logger interface will again be used by the Trainer through Lite, and we don't know whether there will be a "default_root_dir" or not. 

Btw, subdir gets concatenated after the version. A full path would look like this

`root_dir/name/version/sub_dir`

not

`root_dir/name/sub_dir/version`

File: src/lightning_fabric/fabric.py (hunk 8 above)
Comment (lightningforever) [reply src/lightning_fabric/fabric.py:166]: This can already be accessed like so through an index:

```py
fabric.loggers[2]
```

and 

```py
fabric.logger == fabric.loggers[0]
```

Or did you mean something else?

File: src/lightning_fabric/loggers/tensorboard.py (hunk 6 above)
Comment (lightningforever) [reply src/lightning_fabric/loggers/tensorboard.py:44]: I think we want to eventually. In Lite however we currently have more freedom to make minor changes without triggering us to do immediately do refactors that touch many areas of Trainer.

Comment (tchaton) [reply src/lightning_fabric/loggers/tensorboard.py:44]: Yes, let's make them as we have the freedom to do so.

**Review by tchaton — APPROVED** [review]
LGMT !

File: src/lightning_fabric/loggers/tensorboard.py (hunk 5 above)
Comment (justusschock) [reply src/lightning_fabric/loggers/tensorboard.py:84]: I know, but IMO we should simplify as much as possible here. If I see a logger I expect there to be an option with one path and I'd have it log exactly at that path. TBH this is a change I would love to also make for the trainer (once we can without backwards compatibility).

Comment (lightningforever) [reply src/lightning_fabric/loggers/tensorboard.py:84]: I like that idea, I can explore it, but would like to do that in a separate PR because it would change all tests. Since the implementation is copied from PL, I'd like to keep the changes in this initial PR as minimal as possible.

File: src/lightning_fabric/loggers/logger.py (hunk 4 above)
Comment (carmocca) [reply src/lightning_fabric/loggers/logger.py:65]: Regardless of what `fabric.log()` supports, the logger interface don't need to have a matching name. This is how it is in PL at the moment.

File: src/lightning_fabric/loggers/tensorboard.py (hunk 6 above)
Comment (carmocca) [reply src/lightning_fabric/loggers/tensorboard.py:44]: I don't think we are planning any changes to the logger objects themselves. Do you have any in mind? And any differences could be resolved by inheritance if necessary later.

Comment (lightningforever) [reply src/lightning_fabric/loggers/tensorboard.py:44]: The changes are listed in the PR summary.
Also this: https://github.com/Lightning-AI/lightning/pull/16121#discussion_r1054375539

File: src/lightning_fabric/loggers/logger.py (hunk 4 above)
Comment (lightningforever) [reply src/lightning_fabric/loggers/logger.py:65]: Not planning to change it in this PR. We should continue the discussion elsewhere, also in the context of what self.log should do and where in 2.0. Resolving

File: docs/source-pytorch/fabric/fabric.rst [hunk 9]
```diff
@@ -589,3 +613,22 @@ For single-device strategies, it is a no-op. There are strategies that don't sup
 - xla
 
 For these, the context manager falls back to a no-op and emits a warning.
+
+
+log and log_dict
```
Comment (carmocca) [inline docs/source-pytorch/fabric/fabric.rst:618]: This addition and the one above look repeated. They basically say the same thing. Keep one? Merge them?

File: docs/source-pytorch/fabric/fabric.rst [hunk 10]
```diff
@@ -589,3 +613,22 @@ For single-device strategies, it is a no-op. There are strategies that don't sup
...
+    # Set the logger in Fabric
+    fabric = Fabric(loggers=TensorBoardLogger(...))
+
+    # Anywhere in your training loop or model:
+    fabric.log("loss", loss)
+
+    # Or send multiple metrics at once:
+    fabric.log_dict({"loss": loss, "accuracy": acc})
+
+If no loggers are given to Fabric (default), `log` and `log_dict` won't do anything.
```
Comment (carmocca) [inline docs/source-pytorch/fabric/fabric.rst:634]: Can you add two lines of pseudocode showing that `fabric.log` basically routes to `logger.log_*`? This way it's clear to the users that they could also use the logger themselves.

File: src/pytorch_lightning/core/module.py [hunk 11]
```diff
@@ -291,12 +291,12 @@ def truncated_bptt_steps(self, truncated_bptt_steps: int) -> None:
         self._truncated_bptt_steps = truncated_bptt_steps
 
     @property
-    def logger(self) -> Optional[Logger]:
+    def logger(self) -> Optional[Union[Logger, TensorBoardLogger]]:
```
Comment (carmocca) [inline src/pytorch_lightning/core/module.py:294]: Did you make this change because of mypy? This shouldn't be necessary, no?

File: src/pytorch_lightning/loggers/tensorboard.py [hunk 12]
```diff
@@ -19,37 +19,28 @@
...
 from lightning_fabric.utilities.types import _PATH
+from pytorch_lightning.callbacks import ModelCheckpoint
 from pytorch_lightning.core.saving import save_hparams_to_yaml
-from pytorch_lightning.loggers.logger import Logger, rank_zero_experiment
 from pytorch_lightning.utilities.imports import _OMEGACONF_AVAILABLE
 from pytorch_lightning.utilities.rank_zero import rank_zero_only, rank_zero_warn
 
 log = logging.getLogger(__name__)
 
 _TENSORBOARD_AVAILABLE = RequirementCache("tensorboard")
```
Comment (carmocca) [inline src/pytorch_lightning/loggers/tensorboard.py:38]: This can be imported from Fabric

File: src/pytorch_lightning/loggers/logger.py [hunk 13]
```diff
@@ -66,84 +39,12 @@ def after_save_checkpoint(self, checkpoint_callback: ModelCheckpoint) -> None:
...
-
-    def __getitem__(self, idx: int) -> "DummyExperiment":
-        # enables self.logger.experiment[0].add_image(...)
-        return self
-
-    def __setitem__(self, *args: Any, **kwargs: Any) -> None:
-        pass
-
 
 class DummyLogger(Logger):
```
Comment (carmocca) [inline src/pytorch_lightning/loggers/logger.py:49]: Is this not relevant for Fabric?

File: tests/tests_fabric/conftest.py [hunk 14]
```diff
@@ -54,6 +54,7 @@ def restore_env_variables():
         "RANK",  # set by DeepSpeed
         "POPLAR_ENGINE_OPTIONS",  # set by IPUStrategy
         "CUDA_MODULE_LOADING",  # leaked since PyTorch 1.13
+        "CRC32C_SW_MODE",  # leaked by tensorboardX
```
Comment (carmocca) [inline tests/tests_fabric/conftest.py:57]: Fabric tests with tensorboard. Either the comment is not correct or this shouldn't be necessary

File: tests/tests_fabric/loggers/test_tensorboard.py [hunk 15]
```diff
@@ -0,0 +1,229 @@
...
+from unittest import mock
+from unittest.mock import Mock
+
+import numpy as np
+import pytest
+import torch
+from tests_fabric.test_fabric import BoringModel
+
+from lightning_fabric.loggers import TensorBoardLogger
+from lightning_fabric.loggers.tensorboard import _TENSORBOARD_AVAILABLE
```
Comment (carmocca) [inline tests/tests_fabric/loggers/test_tensorboard.py:26]: Do you plan to remove any TB/logger tests from PL?

File: src/pytorch_lightning/core/module.py (hunk 11 above)
Comment (lightningforever) [reply src/pytorch_lightning/core/module.py:294]: Logger here is the Logger base class from PyTorch Lightning
TensorBoardLogger here is subclassing FabricTensorBoardLogger which is subclassing the base class in Fabric

So these don't have the same base class, and therefore need to be included for typing separately. This is a consequence of your request to reuse code through the inheritance.

File: src/pytorch_lightning/loggers/logger.py (hunk 13 above)
Comment (lightningforever) [reply src/pytorch_lightning/loggers/logger.py:49]: Not needed for now. If it's needed we will move it. Note: If I had moved it, another version of @carmocca would have said: "This isn't used anywhere in Fabric, so we should not have it there" :))

File: tests/tests_fabric/conftest.py (hunk 14 above)
Comment (lightningforever) [reply tests/tests_fabric/conftest.py]: What does it matter. It is the same as in pytorch's version of conftest. 
Whether the tests run one way or another depends on which package is installed. Either you have TB installed, or you have TBX installed or both, or none. The tests should run/skip accordingly no matter which configuration.

File: docs/source-pytorch/fabric/fabric.rst (hunk 9 above)
Comment (lightningforever) [reply docs/source-pytorch/fabric/fabric.rst:618]: The above is for the loggers Fabric argument, the one here if for the method(s). I want people to find this information regardless of where they look. This is very brief, there will be a separate guide page that goes into more details.

Comment (lightningforever) [reply docs/source-pytorch/fabric/fabric.rst:618]: Organization will mainly happen in https://github.com/Lightning-AI/lightning/pull/16291 and following PRs.

File: docs/source-pytorch/fabric/fabric.rst (hunk 10 above)
Comment (lightningforever) [reply docs/source-pytorch/fabric/fabric.rst]: done

File: tests/tests_fabric/loggers/test_tensorboard.py (hunk 15 above)
Comment (lightningforever) [reply tests/tests_fabric/loggers/test_tensorboard.py:26]: I think it is too soon now. The loggers in Fabric may change in ways that require us to do overrides again on the PL side (yes I am still against sharing the code through the subclass)

File: src/pytorch_lightning/loggers/logger.py (hunk 13 above)
Comment (carmocca) [reply src/pytorch_lightning/loggers/logger.py:49]: You did the correct thing by not moving it. I was just asking if there were any future plans to use it in Fabric. I guess not because there's no concept of fast dev run

File: src/pytorch_lightning/core/module.py (hunk 11 above)
Comment (carmocca) [reply src/pytorch_lightning/core/module.py]: This is passing mypy locally for me: [`2e90ddf` (#16121)](https://github.com/Lightning-AI/lightning/pull/16121/commits/2e90ddf7d0738bb9f24c268e93a5330dd3ff4c2f)

It just changes the base import to Fabric's. Do you see any issues with this?

**Review by carmocca — APPROVED** [review]

File: src/pytorch_lightning/loggers/logger.py (hunk 13 above)
Comment (lightningforever) [reply src/pytorch_lightning/loggers/logger.py:49]: Yes. One could build a trainer with fabric and have a fast_dev_run feature. In this case, this dummy logger would be a mere utility piece in Fabric.
